core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -308,11 +325,115 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
         return true;
     }
 
-    private RestHandler getHandler(RestRequest request) {
+    void tryAllHandlers(final RestRequest request, final RestChannel channel, final ThreadContext threadContext) throws Exception {
+        for (String key : headersToCopy) {
+            String httpHeader = request.header(key);
+            if (httpHeader != null) {
+                threadContext.putHeader(key, httpHeader);
+            }
+        }
+        // Between retrieving the correct path, we need to reset the parameters,
+        // otherwise parameters are parsed out of the URI that aren't actually handled.
+        final Map<String, String> originalParams = new HashMap<>(request.params());
+
+        // Request execution flag
+        boolean requestHandled = false;
+
+        /*
+         * First try handlers mapped to explicit paths only.
+         */
+        if (requestHandled == false) {
+            final RestHandler handler = getHandler(request, TrieMatchingMode.EXPLICIT_NODES_ONLY);
```

Comment:
I do think that as a subsequent change we should try to make PathTrie and RestHandler a little more decoupled when it comes to mutation and param parsing, I think it would be too complex for this (already complex) PR though